File: FruitWholesale.Client/src/app/features/stock/stock.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { PaginatedList, PaginationRequest } from '../../core/models/common.model';
import { CurrentStock, StockAdjustment, StockLedgerEntry } from '../../core/models/stock.model';
import { toHttpParams } from '../../core/utils/http-params.util';

@Injectable({ providedIn: 'root' })
export class StockService {
  private readonly baseUrl = `${environment.apiUrl}/stock`;

  constructor(private readonly http: HttpClient) {}

  getCurrentStock(): Observable<CurrentStock[]> {
    return this.http.get<CurrentStock[]>(`${this.baseUrl}/current`);
  }

  getStockLedger(fruitId: number, request: PaginationRequest, fromDate?: string | null, toDate?: string | null) {
    return this.http.get<PaginatedList<StockLedgerEntry>>(`${this.baseUrl}/ledger/${fruitId}`, {
      params: toHttpParams({ ...request, fromDate, toDate })
    });
  }

  applyAdjustment(dto: StockAdjustment): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/adjustment`, dto);
  }
}

File: FruitWholesale.Mobile/lib/features/stock/stock_service.dart
import '../../core/api/api_client.dart';
import '../../core/models/paginated_list.dart';
import 'stock_models.dart';

class StockService {
  final ApiClient _api;

  StockService(this._api);

  Future<List<CurrentStock>> getCurrentStock() async {
    final json = await _api.get('/stock/current') as List;
    return json.map((e) => CurrentStock.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<PaginatedList<StockLedgerEntry>> getStockLedger(int fruitId, {int pageNumber = 1}) async {
    final json = await _api.get('/stock/ledger/$fruitId', query: {'pageNumber': pageNumber, 'pageSize': 20});
    return PaginatedList.fromJson(json as Map<String, dynamic>, StockLedgerEntry.fromJson);
  }

  Future<void> applyAdjustment({required int fruitId, required double quantity, required bool isIncrease, required String narration}) async {
    await _api.post('/stock/adjustment', body: {
      'fruitID': fruitId,
      'quantity': quantity,
      'isIncrease': isIncrease,
      'narration': narration,
    });
  }
}
